tool/gravity/cli/logging.go

```diff
@@ -0,0 +1,109 @@
+	"fmt"
+	"log/syslog"
+	"strings"
+
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+)
+
+// LogCLIRunning writes the running cmd as a log entry into the system journal
+// with the gravity-cli tag.
+func LogCLIRunning(cmd string) {
+	entry := fmt.Sprintf("[RUNNING]: %s", cmd)
+	if err := utils.SyslogWrite(syslog.LOG_INFO, entry, constants.GravityCLITag); err != nil {
+		log.WithError(err).Warn("Failed to write to system logs.")
+	}
+}
+
+// LogCLICompleted writes the completed cmd as a log entry into the system journal
+// with the gravity-cli tag. Failed commands will be logged with the returned
+// error.
+func LogCLICompleted(cmd string, err error) {
+	var entry string
+	if err != nil {
+		entry = fmt.Sprintf("[FAILURE]: %s: [ERROR]: %s", cmd, trace.UserMessage(err))
+	} else {
+		entry = fmt.Sprintf("[SUCCESS]: %s", cmd)
+	}
+	if err := utils.SyslogWrite(syslog.LOG_INFO, entry, constants.GravityCLITag); err != nil {
+		log.WithError(err).Warn("Failed to write to system logs.")
+	}
+}
+
+// SanitizeCmd removes potentially sensitive data from the cmdString.
+func SanitizeCmd(g *Application, cmd, cmdString string) string {
+	switch cmd {
+	case g.InstallCmd.FullCommand():
+		if *g.InstallCmd.Token != "" {
+			cmdString = Redact(cmdString, *g.InstallCmd.Token)
+		}
+	case g.JoinCmd.FullCommand():
+		if *g.JoinCmd.Token != "" {
+			cmdString = Redact(cmdString, *g.JoinCmd.Token)
+		}
+	case g.AutoJoinCmd.FullCommand():
+		if *g.AutoJoinCmd.Token != "" {
+			cmdString = Redact(cmdString, *g.AutoJoinCmd.Token)
+		}
+	case g.WizardCmd.FullCommand():
+		if *g.WizardCmd.Token != "" {
+			cmdString = Redact(cmdString, *g.WizardCmd.Token)
+		}
+	case g.OpsAgentCmd.FullCommand():
+		if *g.OpsAgentCmd.Token != "" {
+			cmdString = Redact(cmdString, *g.OpsAgentCmd.Token)
+		}
+	case g.APIKeyDeleteCmd.FullCommand():
+		if *g.APIKeyDeleteCmd.Token != "" {
+			cmdString = Redact(cmdString, *g.APIKeyDeleteCmd.Token)
+		}
+	case g.AppInstallCmd.FullCommand():
+		if *g.AppInstallCmd.RegistryPassword != "" {
+			cmdString = Redact(cmdString, *g.AppInstallCmd.RegistryPassword)
+		}
+	case g.AppUpgradeCmd.FullCommand():
+		if *g.AppUpgradeCmd.RegistryPassword != "" {
+			cmdString = Redact(cmdString, *g.AppUpgradeCmd.RegistryPassword)
+		}
+	case g.AppSyncCmd.FullCommand():
+		if *g.AppSyncCmd.RegistryPassword != "" {
+			cmdString = Redact(cmdString, *g.AppSyncCmd.RegistryPassword)
+		}
+	case g.OpsConnectCmd.FullCommand():
+		if *g.OpsConnectCmd.Password != "" {
+			cmdString = Redact(cmdString, *g.OpsConnectCmd.Password)
+		}
+	case g.UserCreateCmd.FullCommand():
+		if *g.UserCreateCmd.Password != "" {
+			cmdString = Redact(cmdString, *g.UserCreateCmd.Password)
+		}
+	}
+
+	return cmdString
+}
+
+// Redact replaces any instances of value found in cmd with the redacted string.
+func Redact(cmd, value string) string {
+	return strings.ReplaceAll(cmd, value, constants.Redacted)
```

Comment:
Although unlikely, such context-free application of text replacement can replace unrelated portions if two flags happen to have the same value.
Also, we've already solved a similar problem when moving the install/expand functionality to a server running as a systemd service where we had to manipulate the existing command line and add/remove/modify parameters on the fly - this is implemented as [CommandArgs](https://pkg.go.dev/github.com/gravitational/gravity@v0.0.0-20200603182942-2a15a2722622/lib/utils/cli?tab=doc#CommandArgs):

For this to work, we'll need to tweak `CommandArgs` to support flag replacement but it is trivial:

Add the corresponding attribute (`FlagsToReplace`):
```go
// CommandArgs manipulates command-line arguments.
type CommandArgs struct {
	// Parser is used to parse provided command-line.
	Parser ArgsParser
	// FlagsToAdd is a list of additional flags to add to resulting command.
	FlagsToAdd []Flag
	// FlagsToReplace is a list of flags to replace in the command.
	// A flag is only replaced if it has been provided
	FlagsToReplace []Flag
	// FlagsToRemove is a list of flags to omit from the resulting command.
	FlagsToRemove []string
}
```
update `addRemoveFlags` to take that into account:
```go
func (r *CommandArgs) addRemoveFlags(ctx *kingpin.ParseContext) (flags []Flag) {
	var args []Flag
	seen := make(map[string]struct{})
	for _, el := range ctx.Elements {
		switch c := el.Clause.(type) {
		case *kingpin.ArgClause:
			if utils.StringInSlice(r.FlagsToRemove, c.Model().Name) {
				// Remove the positional argument
				continue
			}
			seen[c.Model().Name] = struct{}{}
			args = append(args, NewArg(c.Model().Name, *el.Value))
		case *kingpin.FlagClause:
			if utils.StringInSlice(r.FlagsToRemove, c.Model().Name) {
				// Remove the flag
				continue
			}
			seen[c.Model().Name] = struct{}{}
			if flagInFlags(r.FlagsToReplace, c.Model().Name) {
				// Skip the flag after it's been marked as seen
				continue
			}
			if _, ok := c.Model().Value.(boolCmdlineFlag); ok {
				flags = append(flags, newBoolFlag(c.Model().Name, *el.Value))
			} else {
				flags = append(flags, NewFlag(c.Model().Name, *el.Value))
			}
		}
	}
	for _, flag := range r.FlagsToAdd {
		if arg, ok := flag.(arg); ok {
			if _, exists := seen[arg.name]; !exists {
				args = append(args, arg)
			}
			continue
		}
		if _, exists := seen[flag.Name()]; !exists {
			flags = append(flags, flag)
		}
	}
	for _, flag := range r.FlagsToReplace {
		if _, exists := seen[flag.Name()]; exists {
			flags = append(flags, flag)
		}
	}
	// Return the new command line with positional arguments following non-positional flags
	return append(flags, args...)
}

// ...

func flagInFlags(flags []Flag, flag string) bool {
	for _, f := range flags {
		if f.Name() == flag {
			return true
		}
	}
	return false
}
```

and then just use it as following:

```go
// This is just for illustration - the function already exists in tool/gravity/cli
func parseArgs(args []string) (*kingpin.ParseContext, error) {
	app := kingpin.New("gravity", "")
	app.Terminate(func(int) {})
	app.Writer(ioutil.Discard)
	return RegisterCommands(app).ParseContext(args)
}

func redactCmd(args ...string) (cmd []string, err error) {
	commandArgs := cli.CommandArgs{
		Parser: cli.ArgsParserFunc(parseArgs),
		FlagsToReplace: []Flag{
			cli.NewFlag("token", defaults.Redacted),
			cli.NewFlag("registry-password", defaults.Redacted),
			cli.NewFlag("password", defaults.Redacted),
		},
	}
	args, err = commandArgs.Update(args)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return append([]string{utils.Exe.Path}, args...), nil
}
```

Reply:
@a-palchikov Thanks for pointing this out. This will help simplify the implementation quite a bit.